File: platform/polyglot/fetch_http_cache_response/py/fetch_http_cache_response/token_manager.py
"""Auth token refresh manager with pluggable strategies (Story 2)."""

from __future__ import annotations

import asyncio
import time
from typing import Any, Awaitable, Callable, Protocol

from auth_encoding import encode_auth

from .exceptions import FetchCacheAuthError
from .logger import create as create_logger
from .types import AuthRefreshConfig

logger = create_logger(__name__)


# ─── Strategy Protocol ────────────────────────────────────────────────────────


class TokenStrategy(Protocol):
    async def get_token(self) -> str: ...
    def is_expired(self) -> bool: ...


# ─── Built-in Strategies ─────────────────────────────────────────────────────


class StaticTokenStrategy:
    """Never refreshes, returns static token."""

    def __init__(self, token: str):
        self._token = token

    async def get_token(self) -> str:
        return self._token

    def is_expired(self) -> bool:
        return False


class CallableTokenStrategy:
    """Wraps user-provided async function with TTL-based expiration."""

    def __init__(
        self,
        refresh_fn: Callable[[], Awaitable[str]],
        refresh_interval_seconds: float = 1200.0,
    ):
        self._refresh_fn = refresh_fn
        self._refresh_interval = refresh_interval_seconds
        self._token: str | None = None
        self._expires_at: float = 0.0
        self._lock = asyncio.Lock()

    async def get_token(self) -> str:
        if not self.is_expired() and self._token is not None:
            return self._token
        async with self._lock:
            # Double-check after acquiring lock
            if not self.is_expired() and self._token is not None:
                return self._token
            return await self._refresh()

    def is_expired(self) -> bool:
        return time.monotonic() >= self._expires_at

    async def _refresh(self) -> str:
        try:
            self._token = await self._refresh_fn()
            self._expires_at = time.monotonic() + self._refresh_interval
            logger.debug(f"token refreshed, expires in {self._refresh_interval}s")
            return self._token
        except Exception as e:
            raise FetchCacheAuthError(
                f"Token refresh failed: {e}", cause=e
            ) from e


class ComputedTokenStrategy:
    """Wraps overwrite_from_context {{fn:...}} resolver."""

    def __init__(
        self,
        resolver_name: str,
        resolve_fn: Callable[[str], Awaitable[str]] | None = None,
        refresh_interval_seconds: float = 1200.0,
    ):
        self._resolver_name = resolver_name
        self._resolve_fn = resolve_fn
        self._refresh_interval = refresh_interval_seconds
        self._token: str | None = None
        self._expires_at: float = 0.0
        self._lock = asyncio.Lock()

    async def get_token(self) -> str:
        if not self.is_expired() and self._token is not None:
            return self._token
        async with self._lock:
            if not self.is_expired() and self._token is not None:
                return self._token
            return await self._refresh()

    def is_expired(self) -> bool:
        return time.monotonic() >= self._expires_at

    async def _refresh(self) -> str:
        if self._resolve_fn is None:
            raise FetchCacheAuthError(
                f"No resolve_fn provided for computed token '{self._resolver_name}'"
            )
        try:
            self._token = await self._resolve_fn(self._resolver_name)
            self._expires_at = time.monotonic() + self._refresh_interval
            logger.debug(f"computed token '{self._resolver_name}' refreshed")
            return self._token
        except Exception as e:
            raise FetchCacheAuthError(
                f"Computed token resolution failed for '{self._resolver_name}': {e}",
                cause=e,
            ) from e


# ─── Token Refresh Manager ───────────────────────────────────────────────────


class TokenRefreshManager:
    """Manages token lifecycle with pluggable strategy."""

    def __init__(self, strategy: TokenStrategy, auth_config: AuthRefreshConfig):
        self._strategy = strategy
        self._auth_config = auth_config

    async def get_token(self) -> str:
        return await self._strategy.get_token()

    def is_expired(self) -> bool:
        return self._strategy.is_expired()

    async def build_auth_headers(self) -> dict[str, str]:
        """Build HTTP auth headers using current token."""
        token = await self.get_token()
        auth_type = self._auth_config.auth_type

        # Custom header name override (e.g., X-Figma-Token)
        if self._auth_config.api_auth_header_name:
            return {self._auth_config.api_auth_header_name: token}

        # Use auth_encoding for standard auth types
        credentials: dict[str, Any] = {"token": token}
        try:
            return encode_auth(auth_type, credentials)
        except Exception as e:
            raise FetchCacheAuthError(
                f"Failed to encode auth headers for type '{auth_type}': {e}",
                cause=e,
            ) from e


# ─── Factory ──────────────────────────────────────────────────────────────────


def create_token_strategy(auth_config: AuthRefreshConfig) -> TokenStrategy:
    """Create appropriate token strategy from config."""
    if auth_config.refresh_fn is not None:
        return CallableTokenStrategy(
            refresh_fn=auth_config.refresh_fn,
            refresh_interval_seconds=auth_config.refresh_interval_seconds,
        )
    if auth_config.auth_token_resolver is not None:
        return ComputedTokenStrategy(
            resolver_name=auth_config.auth_token_resolver,
            refresh_interval_seconds=auth_config.refresh_interval_seconds,
        )
    if auth_config.auth_token is not None:
        return StaticTokenStrategy(token=auth_config.auth_token)

    raise FetchCacheAuthError(
        "AuthRefreshConfig must provide one of: auth_token, auth_token_resolver, or refresh_fn"
    )


def create_token_manager(auth_config: AuthRefreshConfig) -> TokenRefreshManager:
    """Create a TokenRefreshManager from config."""
    strategy = create_token_strategy(auth_config)
    return TokenRefreshManager(strategy=strategy, auth_config=auth_config)

File: platform/polyglot/fetch_http_cache_response/mjs/src/token-manager.ts
/**
 * Auth token refresh manager with pluggable strategies (Story 2).
 */

import { FetchCacheAuthError } from "./exceptions.js";
import { createLogger } from "./logger.js";
import type { ResolvedAuthConfig } from "./types.js";

const logger = createLogger("token-manager");

// ─── Strategy Interface ──────────────────────────────────────────────────────

export interface TokenStrategy {
  getToken(): Promise<string>;
  isExpired(): boolean;
}

// ─── Built-in Strategies ─────────────────────────────────────────────────────

export class StaticTokenStrategy implements TokenStrategy {
  #token: string;

  constructor(token: string) {
    this.#token = token;
  }

  async getToken(): Promise<string> {
    return this.#token;
  }

  isExpired(): boolean {
    return false;
  }
}

export class CallableTokenStrategy implements TokenStrategy {
  #refreshFn: () => Promise<string>;
  #refreshInterval: number;
  #token: string | null = null;
  #expiresAt = 0;
  #refreshPromise: Promise<string> | null = null;

  constructor(refreshFn: () => Promise<string>, refreshIntervalSeconds = 1200) {
    this.#refreshFn = refreshFn;
    this.#refreshInterval = refreshIntervalSeconds;
  }

  async getToken(): Promise<string> {
    if (!this.isExpired() && this.#token !== null) {
      return this.#token;
    }
    // Mutex via single promise (async-safe)
    if (this.#refreshPromise === null) {
      this.#refreshPromise = this.#refresh();
    }
    try {
      return await this.#refreshPromise;
    } finally {
      this.#refreshPromise = null;
    }
  }

  isExpired(): boolean {
    return performance.now() / 1000 >= this.#expiresAt;
  }

  async #refresh(): Promise<string> {
    try {
      this.#token = await this.#refreshFn();
      this.#expiresAt = performance.now() / 1000 + this.#refreshInterval;
      logger.debug(`token refreshed, expires in ${this.#refreshInterval}s`);
      return this.#token;
    } catch (e) {
      throw new FetchCacheAuthError(
        `Token refresh failed: ${e}`,
        e instanceof Error ? e : undefined,
      );
    }
  }
}

export class ComputedTokenStrategy implements TokenStrategy {
  #resolverName: string;
  #resolveFn: ((name: string) => Promise<string>) | null;
  #refreshInterval: number;
  #token: string | null = null;
  #expiresAt = 0;
  #refreshPromise: Promise<string> | null = null;

  constructor(
    resolverName: string,
    resolveFn: ((name: string) => Promise<string>) | null = null,
    refreshIntervalSeconds = 1200,
  ) {
    this.#resolverName = resolverName;
    this.#resolveFn = resolveFn;
    this.#refreshInterval = refreshIntervalSeconds;
  }

  async getToken(): Promise<string> {
    if (!this.isExpired() && this.#token !== null) {
      return this.#token;
    }
    if (this.#refreshPromise === null) {
      this.#refreshPromise = this.#refresh();
    }
    try {
      return await this.#refreshPromise;
    } finally {
      this.#refreshPromise = null;
    }
  }

  isExpired(): boolean {
    return performance.now() / 1000 >= this.#expiresAt;
  }

  async #refresh(): Promise<string> {
    if (this.#resolveFn === null) {
      throw new FetchCacheAuthError(
        `No resolveFn provided for computed token '${this.#resolverName}'`,
      );
    }
    try {
      this.#token = await this.#resolveFn(this.#resolverName);
      this.#expiresAt = performance.now() / 1000 + this.#refreshInterval;
      logger.debug(`computed token '${this.#resolverName}' refreshed`);
      return this.#token;
    } catch (e) {
      throw new FetchCacheAuthError(
        `Computed token resolution failed for '${this.#resolverName}': ${e}`,
        e instanceof Error ? e : undefined,
      );
    }
  }
}

// ─── Token Refresh Manager ──────────────────────────────────────────────────

export class TokenRefreshManager {
  #strategy: TokenStrategy;
  #authConfig: ResolvedAuthConfig;

  constructor(strategy: TokenStrategy, authConfig: ResolvedAuthConfig) {
    this.#strategy = strategy;
    this.#authConfig = authConfig;
  }

  async getToken(): Promise<string> {
    return this.#strategy.getToken();
  }

  isExpired(): boolean {
    return this.#strategy.isExpired();
  }

  async buildAuthHeaders(): Promise<Record<string, string>> {
    const token = await this.getToken();

    // Custom header name override
    if (this.#authConfig.apiAuthHeaderName) {
      return { [this.#authConfig.apiAuthHeaderName]: token };
    }

    // Standard auth encoding
    const authType = this.#authConfig.authType;
    try {
      const { encodeAuth } = await import("@internal/auth-encoding") as any;
      return encodeAuth(authType, { token });
    } catch (e) {
      // Fallback: simple Bearer header
      if (authType === "bearer") {
        return { Authorization: `Bearer ${token}` };
      }
      throw new FetchCacheAuthError(
        `Failed to encode auth headers for type '${authType}': ${e}`,
        e instanceof Error ? e : undefined,
      );
    }
  }
}

// ─── Factory ────────────────────────────────────────────────────────────────

export function createTokenStrategy(authConfig: ResolvedAuthConfig): TokenStrategy {
  if (authConfig.refreshFn !== null) {
    return new CallableTokenStrategy(
      authConfig.refreshFn,
      authConfig.refreshIntervalSeconds,
    );
  }
  if (authConfig.authTokenResolver !== null) {
    return new ComputedTokenStrategy(
      authConfig.authTokenResolver,
      null,
      authConfig.refreshIntervalSeconds,
    );
  }
  if (authConfig.authToken !== null) {
    return new StaticTokenStrategy(authConfig.authToken);
  }
  throw new FetchCacheAuthError(
    "AuthRefreshConfig must provide one of: authToken, authTokenResolver, or refreshFn",
  );
}

export function createTokenManager(authConfig: ResolvedAuthConfig): TokenRefreshManager {
  const strategy = createTokenStrategy(authConfig);
  return new TokenRefreshManager(strategy, authConfig);
}
